contrib/natss/pkg/dispatcher/dispatcher/dispatcher.go

```diff
@@ -37,26 +38,26 @@ type SubscriptionsSupervisor struct {
 
 	receiver   *provisioners.MessageReceiver
 	dispatcher *provisioners.MessageDispatcher
-
-	natssConn *stan.Conn
-
-	subscriptionsMux sync.Mutex
-	subscriptions    map[provisioners.ChannelReference]map[subscriptionReference]*stan.Subscription
+	connect    chan struct{}
+	natssURL   string
+	// subscriptionsMux is used to protect updates of subscriptions, natssConn and natssConnInProgress
+	// Dispatcher now dynamically reconnects to NATSS and subscriptionsMux mutex is used
+	// to protect the transition from not connected to connected states.
+	subscriptionsMux    sync.Mutex
+	subscriptions       map[provisioners.ChannelReference]map[subscriptionReference]*stan.Subscription
+	natssConn           *stan.Conn
+	natssConnInProgress bool
 }
 
 // NewDispatcher returns a new SubscriptionsSupervisor.
 func NewDispatcher(natssUrl string, logger *zap.Logger) (*SubscriptionsSupervisor, error) {
 	d := &SubscriptionsSupervisor{
 		logger:        logger,
 		dispatcher:    provisioners.NewMessageDispatcher(logger.Sugar()),
+		connect:       make(chan struct{}),
```

Comment:
I think this is a blocking Channel, should it be non-blocking?

If so, then the various writes to it below can be made into non-blocking writes, because as long as there is at least one item in the channel, a reconnection attempt will occur, so multiple items in the channel aren't needed.